### PR TITLE
feat(node): PR-A autopilot skeleton and mep status

### PR DESCRIPTION
**Pull request by WUAIBING**
## Summary\n- add 
ode/mep_autopilot_config.py with env-based config parsing and validation\n- add 
ode/mep_autopilot_daemon.py scaffold (--status, --once, loop skeleton)\n- add 
ode/mep_status.py convenience status command\n- add focused tests for config validation in 	ests/test_autopilot_config.py\n- document PR-A scaffold commands in APPENDIX.md\n\n## Scope\n- phase A skeleton only\n- no autonomous DM/compute execution yet\n\n## Validation\n- python -m ruff check node/mep_autopilot_config.py node/mep_autopilot_daemon.py node/mep_status.py tests/test_autopilot_config.py\n- python -m pytest tests/test_autopilot_config.py -q -rA\n- python -m node.mep_autopilot_daemon --status\n

### CODE REVIEW

**Comment by WUAIBING** [comment]
## PR Review — Hub Sentinel

Phase A skeleton — overall clean implementation. A few nits:

### Issues

**1. Test assertion never fails (weak test)**
`test_autopilot_config_rejects_min_bounty_above_max` uses:
```python
with pytest.raises(ConfigValidationError, match="cannot be greater"):
```
But the actual code does `if cfg.min_bounty > cfg.max_bounty: raise ConfigValidationError(...)` without any `match=` string. The `match=` parameter makes pytest verify the error message — since there's no message, the test passes regardless. Either add an explicit message string to the raise, or remove the `match=` parameter.

**2. `_validate_url` is dead code**
Defined at line ~38 but never called — the URL validation is done inline in `from_env`. Either use it or remove it.

**3. Negative bounty not validated**
Only `min > max` is checked, but `MEP_MIN_BOUNTY=-5` would be accepted silently. Consider adding:
```python
if cfg.min_bounty < 0 or cfg.max_bounty < 0:
    raise ConfigValidationError("Bounty bounds must be >= 0")
```

**4. `node/__init__.py` check**
`from node.mep_autopilot_config import ...` will fail if `node/__init__.py` doesn't exist or doesn't export. Worth verifying the existing `__init__.py` setup before merging.

### ✅ What's solid
- `frozen=True` dataclass — correct choice for config
- `from_env(Mapping | None)` — testable design
- `timezone_policy` in `as_dict()` — good explicit documentation
- `runtime.note` clearly signals skeleton — no false capabilities
- Comprehensive range validations on task/runtime/spend limits

**Verdict: Approve** ✅ — ready to merge with the nits tracked as follow-up cleanup.

**Review by Moltbot-Sentinel — COMMENTED** [review]
## Hermes Code Review

**Verdict: Approve**

Clean Phase A delivery — config schema, daemon skeleton, status command, tests. The dreaming pattern is clearly visible. Ships exactly what PR #59 promised.

### 💡 Suggestions

- **`mep_autopilot_config.py`** — No `.env` file loading; reads from `os.environ` only. Fine for production (systemd sets env), but for local dev contributors need to `source .env` or `export` manually. Consider a `--env-file` flag in Phase B.
- **`mep_autopilot_daemon.py:82-94`** — The loop is sleep-based, not cron-based. The cron string is stored in config but unused for scheduling. That's fine for Phase A, but Phase B should integrate actual cron or an async scheduler.
- **No state persistence yet** — no `mep_state.json` equivalent. Daemon is stateless. Phase B needs this for: last run timestamp, pending DM queue, failed task history.
- **`mep_status.py`** — Thin 4-line wrapper. Could be a CLI subcommand (`python -m node.mep_autopilot_daemon --status`) instead of separate file. Having it as a convenience shortcut is fine, just noting the duplication.
- **Tests only cover config** — 4 tests, all validation. Good for config module. When Phase B adds execution logic, add tests for `run_tick()` behavior paths (disabled, paused, no_jobs, jobs_planned).

### ✅ Looks Good

- Frozen dataclass with `from_env()` — immutable after creation, testable with mock dicts
- Thorough validation: cron format, URL prefixes, bounty range, safety caps
- Safe defaults: everything `false` by default, `MAX_TOKEN_SPEND_PER_HOUR=1000` (fixes the `***` from PR #59)
- `run_tick()` returns `"jobs_planned_noop"` — honest scaffolding, clean separation of planning vs execution
- `--once` mode perfect for future cron integration

---
*Reviewed by Hermes Agent*

File: node/mep_autopilot_config.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+
+@dataclass(frozen=True)
+class AutopilotConfig:
+    autopilot_enabled: bool
+    idle_earn_enabled: bool
+    dm_sync_enabled: bool
+    compute_sync_enabled: bool
+    autopilot_cron: str
+    autopilot_timezone: str
```
Comment (Moltbot-Sentinel) [inline node/mep_autopilot_config.py:57]: 💡 **No `.env` file support** — `source = env if env is not None else os.environ` only reads from the process environment. For local dev, contributors need to manually `export` every var. Consider adding a `--env-file` CLI flag in Phase B that loads a `.env` file before `from_env()` runs.

File: node/mep_autopilot_daemon.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        print(f"[autopilot] config error: {exc}")
+        return 2
+
+    if args.status:
+        print_status(config)
+        return 0
+
+    if args.once:
+        print(json.dumps(run_tick(config), indent=2, sort_keys=True))
+        return 0
```
Comment (Moltbot-Sentinel) [inline node/mep_autopilot_daemon.py:82]: 💡 **Sleep loop vs cron** — `time.sleep(max(args.sleep_seconds, 0.1))` is a dev placeholder. The config stores `autopilot_cron` but it's never parsed or used for scheduling. Phase B should either: (a) integrate `croniter` for tick scheduling, or (b) document that external cron should call `--once` (like the dreaming pattern).

File: node/mep_autopilot_daemon.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        "mode": "skeleton",
+        "timestamp_utc": _utc_now_iso(),
+        "note": "PR-A scaffold only. No autonomous DM/compute actions are executed yet.",
+    }
+    return snapshot
+
+
+def run_tick(config: AutopilotConfig) -> dict:
+    jobs = []
+    if config.idle_earn_enabled:
```
Comment (Moltbot-Sentinel) [inline node/mep_autopilot_daemon.py:26]: 💡 **No state persistence** — `run_tick()` is stateless. Phase B will need a state file (e.g. `mep_state.json`) to track: `last_tick_ts`, `pending_dm_queue`, `failed_task_ids`, and `ticks_run` counter. The dreaming system's `dream_state.json` is a good template — load at start, save at exit, resume-safe.

File: node/mep_status.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
+from node.mep_autopilot_config import AutopilotConfig, ConfigValidationError
```
Comment (Moltbot-Sentinel) [inline node/mep_status.py:1]: 💡 **Consider consolidation** — this is a 4-line wrapper around `print_status()`. The same output is available via `python -m node.mep_autopilot_daemon --status`. Having both is fine for discoverability, but could confuse contributors about which one to use.

File: tests/test_autopilot_config.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import pytest
```
Comment (Moltbot-Sentinel) [inline tests/test_autopilot_config.py:1]: 💡 **Future test targets** — when Phase B adds execution logic to `run_tick()`, add tests for each decision path: `autopilot_disabled`, `autopilot_paused`, `no_jobs_enabled`, and the actual execution path. The current 4 tests cover config validation well.

**Comment by Moltbot-Sentinel** [comment]
## Review from Moltbot

Solid Phase A scaffold. A few thoughts:

**What works:**
- Clean `ConfigValidationError` with proper error types ✅
- `--status` and `--once` flags are good for debugging and CI ✅
- Tests included from the start ✅
- `build_status_snapshot()` is well-structured for future extension ✅

**Questions/notes:**

1. **Hub connectivity check** — `--status` validates config but does not verify the node is actually registered/connected to the Hub. Consider adding an optional ping to confirm reachability.

2. **Cron validation is shallow** — Only checks field count, not valid ranges (e.g., `hour=25` would pass). Minor but could catch typos early.

3. **Entry point** — How do users run `mep_status`? May need a `__main__.py` or console script entry in `pyproject.toml`.

**Verdict:** Good scaffold, structure will scale well for Phase B/C. Ready to merge. ✅

File: node/mep_autopilot_daemon.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+def build_status_snapshot(config: AutopilotConfig) -> dict:
+    snapshot = config.as_dict()
+    snapshot["status"] = "paused" if config.autopilot_pause else "ready"
+    snapshot["runtime"] = {
+        "mode": "skeleton",
+        "timestamp_utc": _utc_now_iso(),
+        "note": "PR-A scaffold only. No autonomous DM/compute actions are executed yet.",
+    }
+    return snapshot
+
```
Comment (WUAIBING) [inline node/mep_autopilot_daemon.py:22]: **🔍 Hub-Sentinel Review - Phase A Skeleton**

Good scaffold. A few issues I found:

---

### 🔴 Critical: `as_dict()` method missing

`build_status_snapshot()` calls `config.as_dict()` but `AutopilotConfig` is a frozen dataclass with **no `as_dict()` method**. This will raise `AttributeError` at runtime when `--status` is called.

**Fix:** Add `as_dict()` method or use `dataclasses.asdict(config)`.

---

### 🟡 Security: URL validation is incomplete

`_validate_url()` only checks `startswith`. It does not handle:
- Port parsing: `https://127.0.0.1:8080` would pass
- Fragment/params: `https://internal-corp.local#evil`
- IPv6: `https://[::1]`

For a skeleton that will eventually make outbound requests, consider:
- Blocking private IP ranges
- Stripping fragments and query params

---

### 🟡 Bug: `min_bounty` / `max_bounty` no ordering check

Nothing validates that `MEP_MIN_BOUNTY <= MEP_MAX_BOUNTY`. If someone sets `MIN=20, MAX=5`, runtime behavior will be confusing.

---

### 🟢 Nit: `idle_required` is stored but never used

Field is defined and parsed but never referenced in `run_tick()`. Either wire it up in Phase B or remove it from the skeleton.

---

### 🟢 Nit: Loop exception kills daemon silently

If `run_tick()` raises an exception, the daemon exits with no log. Consider wrapping in try/except with backoff.

---

**Overall:** Solid Phase A. The `as_dict()` bug is the only blocker. Rest are Phase B material.

File: node/mep_autopilot_config.py (hunk 1 above)
Comment (WUAIBING) [inline node/mep_autopilot_config.py:57]: **🔍 Hub-Sentinel Review**

### 🔴 Missing: `as_dict()` method

`build_status_snapshot()` in the daemon calls `config.as_dict()` but `AutopilotConfig` is a frozen dataclass without this method. Runtime will crash with `AttributeError`.

**Fix options:**
1. Add `def as_dict(self) -> dict: return dataclasses.asdict(self)` to the class
2. Or import `dataclasses` and call `dataclasses.asdict(config)` in the daemon

---

### 🟡 Add: `min_bounty <= max_bounty` validation

If someone sets `MEP_MIN_BOUNTY=20` and `MEP_MAX_BOUNTY=5`, no error is raised. Add ordering check in `from_env()`:

```python
if cfg.min_bounty > cfg.max_bounty:
    raise ConfigValidationError("MEP_MIN_BOUNTY must be <= MEP_MAX_BOUNTY")
```

---

### 🟢 Consider: Document the 5-field cron requirement in APPENDIX.md

_validate_cron() enforces classic 5-field syntax. Contributors should know this upfront.

File: node/mep_autopilot_daemon.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        jobs.append("idle_earn")
+    if config.dm_sync_enabled:
+        jobs.append("dm_sync")
+    if config.compute_sync_enabled:
+        jobs.append("compute_sync")
+
+    if not config.autopilot_enabled:
+        decision = "autopilot_disabled"
+    elif config.autopilot_pause:
+        decision = "autopilot_paused"
```
Comment (WUAIBING) [inline node/mep_autopilot_daemon.py:36]: ## 🔴 `idle_required` is parsed but never enforced

`AutopilotConfig.idle_required` is parsed in `from_env()` and appears in `as_dict()` output, but `run_tick()` never checks it. If `idle_required=True`, the tick should only run idle-earn jobs when the node is truly idle (no active task). This is a core safety flag — either implement it now or remove it from the config to avoid a false sense of security in Phase B.

**File:** `node/mep_autopilot_config.py` line 27 / `node/mep_autopilot_daemon.py`

**Suggestion:** Add `if config.idle_required and not is_node_idle(): return {"decision": "not_idle", ...}` at the top of `run_tick()`.

Comment (WUAIBING) [inline node/mep_autopilot_daemon.py:36]: ## 🟡 `allowed_models` is stored but never enforced

`AutopilotConfig.allowed_models` is parsed and output in `as_dict()`, but `run_tick()` never checks whether a model is in the allowed list before queuing a job. When Phase B adds actual model selection, this could allow a prohibited model to be used. Consider either:

1. Adding a `is_model_allowed(model: str) -> bool` helper now
2. Documenting that Phase B must filter at job-queue time

**File:** `node/mep_autopilot_config.py` / `node/mep_autopilot_daemon.py`

File: node/mep_autopilot_config.py [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+            "MEP_AUTOPILOT_CRON must contain 5 fields (minute hour day month weekday)"
+        )
+
+
+def _validate_url(url: str, key: str, allowed_prefixes: tuple[str, ...]) -> None:
+    if not url.startswith(allowed_prefixes):
+        raise ConfigValidationError(f"{key} must start with one of {allowed_prefixes}, got: {url!r}")
+
+
+@dataclass(frozen=True)
```
Comment (WUAIBING) [inline node/mep_autopilot_config.py:50]: ## 🟡 `_validate_cron` checks field count but not valid cron values

`_validate_cron()` only verifies the expression has 5 space-separated fields. It does not validate that each field contains a valid cron component (e.g., minute must be 0-59, hour 0-23). A value like `99 * * * *` would pass validation but is invalid.

**Suggestion:** Either add proper range validation per field, or rely on `croniter` in Phase B (which Moltbot already recommends). For Phase A skeleton, a simple range check is sufficient:

```python
minute, hour, dom, month, dow = expr.split()
if not (0 <= int(minute) <= 59): raise ConfigValidationError(...)
```

File: tests/test_autopilot_config.py [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+
+def test_autopilot_config_rejects_invalid_boolean() -> None:
+    with pytest.raises(ConfigValidationError, match="boolean-like"):
+        AutopilotConfig.from_env({"MEP_AUTOPILOT_ENABLED": "maybe"})
+
+
+def test_autopilot_config_rejects_min_bounty_above_max() -> None:
+    with pytest.raises(ConfigValidationError, match="cannot be greater"):
+        AutopilotConfig.from_env({"MEP_MIN_BOUNTY": "10", "MEP_MAX_BOUNTY": "5"})
```
Comment (WUAIBING) [inline tests/test_autopilot_config.py:26]: ## 🟡 Tests miss negative-value validation

`AutopilotConfig.from_env()` validates `max_tasks_per_hour >= 0`, `max_runtime_seconds > 0`, `max_token_spend_per_hour >= 0`, but `test_autopilot_config.py` only tests defaults and the 3 error cases (bad cron, bad bool, bad bounty range). No test covers the boundary conditions for negative/zero values.

**Missing tests:**
- `MEP_MAX_TASKS_PER_HOUR=-1` → should raise `ConfigValidationError`
- `MEP_MAX_RUNTIME_SECONDS=0` → should raise `ConfigValidationError`
- `MEP_MAX_TOKEN_SPEND_PER_HOUR=-100` → should raise `ConfigValidationError`

File: tests/test_autopilot_config.py [hunk 10]
```diff
@@ -0,0 +1,26 @@
+import pytest
+
+from node.mep_autopilot_config import AutopilotConfig, ConfigValidationError
+
+
+def test_autopilot_config_defaults_are_valid() -> None:
+    cfg = AutopilotConfig.from_env({})
+    assert cfg.autopilot_enabled is False
```
Comment (WUAIBING) [inline tests/test_autopilot_config.py:8]: ## 🟢 Nit: `test_autopilot_config_defaults_are_valid` is too loose

The test asserts `cfg.allowed_models` is truthy, which passes for the default `cli-agent,gemini,deepseek`. But if a contributor sets `MEP_ALLOWED_MODELS=""` (empty string), `_parse_csv("")` returns an empty tuple which is falsy — and this case is explicitly validated in `from_env()` with `if not cfg.allowed_models: raise`. There is no test for the empty-string case. Consider adding:

```python
def test_autopilot_config_rejects_empty_allowed_models():
    with pytest.raises(ConfigValidationError, match="at least one model"):
        AutopilotConfig.from_env({"MEP_ALLOWED_MODELS": ""})
```

**File:** `tests/test_autopilot_config.py`

File: APPENDIX.md [hunk 11]
```diff
@@ -37,6 +37,16 @@ Optional:
...
 - Uses `HUB_URL` and `WS_URL` from environment when set
 
+### Autopilot PR-A Skeleton Commands
+
+The Phase A scaffold adds safe status/skeleton commands only (no autonomous DM or compute execution yet):
+
+```bash
+python -m node.mep_status
+python -m node.mep_autopilot_daemon --status
+python -m node.mep_autopilot_daemon --once
```
Comment (WUAIBING) [inline APPENDIX.md:47]: ## 🟡 APPENDIX.md documents commands but not environment variables

The new "Autopilot PR-A Skeleton Commands" section in APPENDIX.md correctly shows how to run the commands, but contributors have no reference for the new env vars introduced by this PR (`MEP_AUTOPILOT_ENABLED`, `MEP_AUTOPILOT_CRON`, `MEP_IDLE_EARN_ENABLED`, `MEP_DM_SYNC_ENABLED`, `MEP_COMPUTE_SYNC_ENABLED`, `MEP_AUTOPILOT_PAUSE`, etc.).

Consider adding an "Autopilot Configuration" subsection in APPENDIX.md mirroring the existing "Node Configuration" pattern, listing each var with its default and purpose. This is especially important since `autopilot_cron` and the safety limits (`max_tasks_per_hour`, `max_runtime_seconds`, etc.) have non-obvious defaults.

**Comment by WUAIBING** [comment]
Addressed review nits in latest push:\n\n- Added explicit non-negative bounty bounds validation (MEP_MIN_BOUNTY / MEP_MAX_BOUNTY must be >= 0)\n- Strengthened test assertions to check exact exception messages for bounty-bound errors\n- Added invalid HUB_URL prefix test to cover URL validation path\n- Added 
ode/__init__.py for explicit package clarity\n\nCI should re-run automatically.

File: node/mep_autopilot_config.py [hunk 12]
```diff
@@ -0,0 +1,141 @@
+import os
```
Comment (Hub-Sentinel) [inline node/mep_autopilot_config.py:1]: ## 🧊 Elsaws Review

Three items worth addressing before or after merge:

### 🔴 `as_dict()` method missing

`build_status_snapshot()` calls `config.as_dict()` but `AutopilotConfig` is a frozen dataclass with no such method. This will `AttributeError` at runtime when `--status` is called.

**Fix:** Add `def as_dict(self) -> dict: return dataclasses.asdict(self)` or call `dataclasses.asdict(config)` in the daemon.

---

### 🟡 `idle_required` is unenforced

`AutopilotConfig.idle_required` is parsed and appears in `as_dict()` output, but `run_tick()` never checks it. If `True`, idle-earn jobs should only run when the node has no active task. Either implement the check now or remove the field to avoid a false sense of security in Phase B.

---

### 🟡 `_validate_cron()` only checks field count

`99 * * * *` passes validation but is invalid. Minutes must be 0-59, hours 0-23. Either add range validation now, or ensure `croniter` in Phase B catches this.

File: node/mep_autopilot_config.py [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+        return float(raw)
+    except ValueError as exc:
+        raise ConfigValidationError(f"{key} must be numeric, got: {raw!r}") from exc
+
+
+def _parse_csv(raw: str) -> tuple[str, ...]:
+    return tuple(item.strip() for item in raw.split(",") if item.strip())
+
+
+def _validate_cron(expr: str) -> None:
```
Comment (WUAIBING) [inline node/mep_autopilot_config.py:37]: **Nit (non-blocking):** `_validate_cron` only checks field count (5 fields) but does not validate value ranges (e.g. minute 0-59, hour 0-23). Acceptable for PR-A scaffold, but worth addressing in Phase B.

> **Suggestion:** Consider `croniter` for proper cron validation in a follow-up.

File: tests/test_autopilot_config.py [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+def test_autopilot_config_rejects_invalid_boolean() -> None:
+    with pytest.raises(ConfigValidationError, match="boolean-like"):
+        AutopilotConfig.from_env({"MEP_AUTOPILOT_ENABLED": "maybe"})
+
+
+def test_autopilot_config_rejects_min_bounty_above_max() -> None:
+    with pytest.raises(ConfigValidationError) as exc:
+        AutopilotConfig.from_env({"MEP_MIN_BOUNTY": "10", "MEP_MAX_BOUNTY": "5"})
+    assert str(exc.value) == "MEP_MIN_BOUNTY cannot be greater than MEP_MAX_BOUNTY"
+
```
Comment (WUAIBING) [inline tests/test_autopilot_config.py:28]: **Minor test gap:** `test_autopilot_config_rejects_negative_bounty_bounds` only tests `min_bounty=-1` but not `max_bounty=-1`. The validation code handles both cases, but the test coverage is asymmetric. Consider adding:

```python
def test_autopilot_config_rejects_negative_max_bounty():
    with pytest.raises(ConfigValidationError, match="must be >= 0"):
        AutopilotConfig.from_env({"MEP_MAX_BOUNTY": "-1"})
```

**Review by Hermes-Agent11 — APPROVED** [review]
## Hermes Review — PR #60: Autopilot Skeleton

**Verdict: APPROVE** ✅

### Post-fix status

The fix commit `8f384f2` addressed the two actionable issues from prior reviews:
- ✅ Negative bounty bounds validation added
- ✅ Hub URL prefix test added
- ✅ Bounty test tightened to exact match

The `as_dict()` method exists on the frozen dataclass (lines 89-115 in config) — the 🔴 from earlier reviews is already resolved in the current diff.

### Remaining items are all Phase B scope

These are intentional gaps for a skeleton PR, not bugs:

| Item | Severity | Why it's OK here |
|---|---|---|
| `idle_required` unenforced | 🟡 | No execution logic yet — nothing to enforce against |
| `allowed_models` unenforced | 🟡 | Same — `run_tick()` is a noop planner |
| Cron validation (field values) | 🟡 | `croniter` dependency belongs in Phase B |
| APPENDIX.md env var docs | 🟡 | Follow-up PR, keeps this PR focused |
| State persistence | 💡 | Phase B design decision |
| Sleep loop vs cron | 💡 | Skeleton intentionally sleep-based |

### What's good

- Frozen dataclass + `from_env()` — testable, immutable, clean
- Safety defaults: everything `false`, explicit opt-in per feature
- `run_tick()` returns `"jobs_planned_noop"` — honest scaffolding
- `--once` mode = perfect cron integration point for Phase B
- Validation is thorough: cron format, URL prefixes, bounty range, safety caps

### Note on review attribution

Moltbot-Sentinel's review is titled "Hermes Code Review" and some WUAIBING comments carry "Hub-Sentinel Review" headers. Minor attribution confusion — worth cleaning up in review bot config.

**Confidence: 0.90** — Clean skeleton, good test coverage for Phase A, all blocking issues resolved in fix commit.
